Trail mark

bianchi-dy opened the pull request:
Referencing #23, changes in this PR:

**Trail properties**

* Trail mark implemented as one area where the path is dicated by adding/subtracting the stroke width for the corresponding coordinate (see figure below) – the path is generated using `d3.line().curve(curveCardinal.tension(0.96))` to get a path similar to Multiline
* Uses `stroke-width` to dictate variable 'stroke'
* Uses `fillOpacity` and `fill` to dictate opacity and color
* `sort` and `close` are enabled

**Trail docs**

* Updated trail docs

**Feedback needed**

* I use a lot of custom logic to implement Trail due to sorting the points and aesthetics together, plus generating the coordinates for the path. Perhaps there are more parts that can be adapted to reuse present code in Path.js (the primary mixin), that are not obvious to me at the moment
* I'm not sure how 'smoothly' the trail looks like on other screens, as the end points of the line are sometimes not rounded due to how the entire path is generated – right now my current implementation is as follows:

1. Convert points to screen coordinates 
2. Get equation of line from current point to next point -> eqn 1 with unit vector 1
3. Find unit vector parallel to unit vector 1 -> unit vector 2 
4. Find points +/- stroke width/2 along unit vector 2 from current point
5. Organize array of path coordinates such that top section is followed by bottom section of path
6. Draw path using curveCardinal

And it ends up looking like this:

<img width="563" alt="screenshot 2019-02-25 at 6 24 26 pm" src="https://user-images.githubusercontent.com/44487900/53353283-54451280-395f-11e9-9fc0-b14871c1cb5d.png">

luucvanderzee commented on docs/.vuepress/components/SymbolMarkDemo.vue:
```diff
@@ -102,7 +98,7 @@ export default {
     fill () {
       if (this.color === 'both' || this.color === 'fill') {
-        return { scale: { scale: 'viridis', variable: 'explanatory' } }
+        return { get: 'explanatory', scale: { type: 'viridis', domain: 'explanatory' } }
```
This syntax is not used anymore... we now use `val` which gets the value directly through the scoped slot

luucvanderzee commented on docs/.vuepress/components/ScatterplotMatrix.vue:
```diff
@@ -0,0 +1,370 @@
+<template>
```
Wow this is a big one lol. This can be improved a lot with the new `vgg-repeat` component, especially after I finish the automatic axis positioning (is currently under construction on my `section-slots` branch).

luucvanderzee replied:
Also, when I run `yarn docs:dev`, I get an compilation error due to this component not having one root element...

bianchi-dy replied:
I think this is leftover from when I was trying to figure out how to merge branches and merged the wrong ones. Thanks for spotting this

luucvanderzee commented on docs/.vuepress/components/Cartesian.vue:
```diff
@@ -0,0 +1,145 @@
+<template>
```
What is this?

bianchi-dy replied:
Leftover from my previous faulty merge attempts, deleting this

luucvanderzee commented on package.json:
```diff
@@ -23,6 +23,7 @@
     "d3-scale-chromatic": "^1.3.3",
     "d3-shape": "^1.2.2",
     "d3-time-format": "^2.1.3",
+    "docs": "^0.0.0-37",
```
?

luucvanderzee commented:
There is a lot stuff from the Axis options branch in here, which we didn't merge into master yet because it is not ready... and a lot of random things like the Scatterplot matrix component, and some old obsolete code seems to have returned (like `mapAesthetics.js`). Is it possible to clean this branch up a little, so that we can merge in just the Trail mark? If it works more or less like the other mark, it should be quite self-contained, so isolating the Trail functionality shouldnt be too hard

bianchi-dy commented:
Okay, I think the commits from today cover what you've mentioned @luucvanderzee 

**Changes**

- Removed leftovers from faulty merge (i.e. obsolete code like mapAesthetics.js, other branch stuff like BaseAxis.js and so on) + random bits unrelated to trail mark
- Added demo to trail mark documentation

luucvanderzee commented on src/components/Marks/Label.vue:
```diff
@@ -1,80 +1,65 @@
 <script>
```
Why did you remove all whitespace from this file?

bianchi-dy replied:
I did a bunch of reverts for the unrelated-to-trail-mark code, this was unintentional

luucvanderzee commented on src/mixins/Marks/Path.js:
```diff
@@ -48,7 +48,7 @@ export default {
 
     strokeWidth: {
-      type: Number,
+      type: [Number, Array, Object, undefined],
```
I am assuming the `Array` is for an array of `strokeWidth` values for the trail mark, is that correct? But why do we need the `Object`? Also, it is probably better to just keep this `type: Number` and overwrite this in the `Trail` mark, so that only the trail mark accepts an array of values.
